server/src/api/socket.js

```diff
@@ -3,6 +3,21 @@ import { getOrCreateContainer, attachContainer } from '../utils/docker.js'
 import Session from '../models/session.js'
 import { defaultUser, genSessionJWT } from '../utils/auth.js'
 
+const sessions = new Map()
+
+function listenToSession(sessionId, event, callback) {
+  const callbacks = sessions.get((sessionId, event)) || []
+  sessions.set((sessionId, event), callbacks.concat(callback))
+}
+
+export function pushToSession(sessionId, event, data) {
+  console.log('Push:', sessionId, event)
+  const callbacks = sessions.get((sessionId, event))
+  if (callbacks) {
+    callbacks.forEach((callback) => callback(data))
+  }
+}
+
```

Comment:
I would like to know why do you use callbacks on this function and how they will funcion

Reply:
The listenToSession is used in [socket.js](https://github.com/lancatlin/linux-odyssey/pull/11/files/c2480044848d48071bd34035170e9734ac28ff1d#diff-b2502043fb2171908d74afb6bd0cf1a5a832dc5450873404f5ec6ac28c4b5753R81). Because the socket instance is initialize in the callback function of io.on('connection'), so it would be hard for other code to send data through that socket. However, using listeners pattern we can reverse the dependencies, so the caller (in this case it's [commandHandler.js](https://github.com/lancatlin/linux-odyssey/pull/11/files/c2480044848d48071bd34035170e9734ac28ff1d#diff-1b7f07d85702483ae70cda0795f22a2d05178075b712dc2424c3741d14789629R25) ) doesn't need to know anything about the real socket underneath, but just push an event and let listeners to handle it.